### ISSUE TITLE
State mismatch for preview3 testnet @140196

### ISSUE THREAD

**Issue by roman-khimov** [nspcc-dev/neo-go#1456]
Transaction 46ccaabc95617de3adb2747419a45187cdef5342f30fdbdb1985a4e0d4884c35 is handled in a different way, but root of the problem seems to be in failed contract deployment:
```
2020-10-02T19:34:24.468+0300    WARN    contract invocation failed      {"tx": "46ccaabc95617de3adb2747419a45187cdef5342f30fdbdb1985a4e0d4884c35", "block": 140196, "error": "error encountered at instruction 80 (SYSCALL): failed to invoke syscall: contract not found"}
```

**Comment by roman-khimov** [comment]
C# application log:
```
{
   "jsonrpc" : "2.0",
   "result" : {
      "txid" : "0x46ccaabc95617de3adb2747419a45187cdef5342f30fdbdb1985a4e0d4884c35",
      "stack" : [
         {
            "value" : "1",
            "type" : "Integer"
         }
      ],
      "vmstate" : "HALT",
      "gasconsumed" : "13799010",
      "trigger" : "Application",
      "notifications" : [
         {
            "contract" : "0x2ff06a43c9f58b452e1e526ade590a0ed97070cb",
            "eventname" : "Transfer",
            "state" : {
               "type" : "Array",
               "value" : [
                  {
                     "value" : "pJfiznpVQgf8F5jTm5uzurossNw=",
                     "type" : "ByteString"
                  },
                  {
                     "value" : "u44ftq67iSikHAj6T2FFqSlgHCM=",
                     "type" : "ByteString"
                  },
                  {
                     "value" : "101",
                     "type" : "Integer"
                  }
               ]
            }
         }
      ]
   },
   "id" : 1
}
```
NeoGo:
```
{
   "result" : {
      "stack" : [],
      "vmstate" : "FAULT",
      "txid" : "0x46ccaabc95617de3adb2747419a45187cdef5342f30fdbdb1985a4e0d4884c35",
      "trigger" : "Application",
      "notifications" : [],
      "gasconsumed" : "1007780"
   },
   "id" : 1,
   "jsonrpc" : "2.0"
}
```

**Comment by roman-khimov** [comment]
Seems to be caused by this:
```
2020-10-02T19:34:11.056+0300    WARN    contract invocation failed      {"tx": "501b643cc3b3da88461aecaed17f90456e92654af046653fb3246ee3873dd20d", "block": 127240, "error": "error encountered at instruction 1268 (SYSCALL): failed to invoke syscall: gas limit exceeded"}
2020-10-02T19:34:11.172+0300    WARN    contract invocation failed      {"tx": "cebd78d703b4c064323a0ab683f241c6b94f9d3098b2910f61759e7a130e294c", "block": 127344, "error": "error encountered at instruction 1267 (SYSCALL): failed to invoke syscall: gas limit exceeded"}
```
C# application logs:
```
{
   "jsonrpc" : "2.0",
   "result" : {
      "notifications" : [],
      "gasconsumed" : "197929760",
      "trigger" : "Application",
      "vmstate" : "HALT",
      "txid" : "0x501b643cc3b3da88461aecaed17f90456e92654af046653fb3246ee3873dd20d",
      "stack" : [
         {
            "value" : "1",
            "type" : "Integer"
         }
      ]
   },
   "id" : 1
}
{
   "jsonrpc" : "2.0",
   "result" : {
      "txid" : "0xcebd78d703b4c064323a0ab683f241c6b94f9d3098b2910f61759e7a130e294c",
      "notifications" : [],
      "trigger" : "Application",
      "stack" : [
         {
            "value" : "1",
            "type" : "Integer"
         }
      ],
      "vmstate" : "HALT",
      "gasconsumed" : "197829760"
   },
   "id" : 1
}
```
VM issue?

**Comment by roman-khimov** [comment]
C#:
```
            AddGas(StoragePrice * (script?.Length ?? 0 + manifest?.Length ?? 0));
```
Go:
```
        if !ic.VM.AddGas(int64(StoragePrice * (len(script) + len(manifestBytes)))) {
                return errGasLimitExceeded
        }
```
But that's where the difference comes from, this transactions has a manifest of 1540 bytes and a script of 1966 bytes. We're counting both and getting 350600000 for the syscall, while C# seems to be counting just 1966 and getting 196600000.

**Comment by fyrchik** [comment]
`??` is right-assotiative [link](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-coalescing-operator), and `+` has higher precedence [link](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/#operator-precedence).
So this is parsed as `script?.Length ?? ((0 + manifest?.Length) ?? 0)` [link](https://dotnetfiddle.net/rD2q5h)

**Comment by roman-khimov** [comment]
There is nothing here to fix for v0.92.0, but we need a kludge for @alexvanin to reliably work with preview3 testnet.

**Comment by fyrchik** [comment]
https://github.com/alexvanin/neo-go/pull/1
